adserver/tasks.py

```diff
@@ -341,90 +363,115 @@ def daily_update_regiontopic(day=None):  # pylint: disable=too-many-branches
+            total_views=Count("country", filter=Q(viewed=True)),
+            total_clicks=Count("country", filter=Q(clicked=True)),
+        )
+        .filter(total_decisions__gt=0)
+        .order_by("-total_decisions")
+        .values(
+            "keywords",
+            "advertisement",
+            "country",
+            "total_decisions",
+            "total_offers",
+            "total_views",
+            "total_clicks",
+        )
+        .iterator()
+    ):
+        if not (values["keywords"] and values["country"]):
+            continue
 
-            if country in us_ca:
-                region = "us-ca"
-            elif country in eu_aus_nz:
-                region = "eu-aus-nz"
-            elif country in wider_apac:
-                region = "wider-apac"
-            elif country in latin_america:
-                region = "latin-america"
-            elif country in africa:
-                region = "africa"
+        keywords = values["keywords"]
+        country = values["country"]
+        ad = values["advertisement"]
+        publisher_keywords = set(keywords)
+
+        topics = set()
+        for keyword in publisher_keywords:
+            if keyword in data_science:
+                topic = "data-science"
+            elif keyword in backend_web:
+                topic = "backend-web"
+            elif keyword in frontend_web:
+                topic = "frontend-web"
+            elif keyword in security_privacy:
+                topic = "security-privacy"
+            elif keyword in devops:
+                topic = "devops"
+            elif keyword in python:
+                topic = "python"
+            elif keyword in blockchain:
+                topic = "blockchain"
+            elif keyword in game_dev:
+                topic = "game-dev"
             else:
-                region = "global"
-
-            # Aggregate data into topic_mapping to save doing queries until we have everything counted
-            # This is important because we can't query on keywords, so we have a lot of records that increment
-            # the total count on the region & topic.
-            for topic in topics:
-                topic_mapping[f"{ad}:{region}:{topic}"] += values["total"]
+                continue
 
-        log.info(
-            "Saving %s RegionTopicImpressions: %s", len(topic_mapping), impression_type
+            topics.add(topic)
+
+        # If nothing gets set as a topic, assign it other
+        if not topics:
+            topics.add("other")
+
+        if country in us_ca:
+            region = "us-ca"
+        elif country in eu_aus_nz:
+            region = "eu-aus-nz"
+        elif country in wider_apac:
+            region = "wider-apac"
+        elif country in latin_america:
+            region = "latin-america"
+        elif country in africa:
+            region = "africa"
+        else:
+            region = "global"
```

Comment:
I think these regions aren't in sync with the other logic around regions, which includes `south-asia`. Another problem with this file not being very DRY (that I introduced).